### DISCUSSION TITLE
What's the status of this repository?

### DISCUSSION THREAD

**Discussion by gijswijs** [node-lightning/node-lightning#1] [Q&A]
I'll come to business right away:
This repository is exactly what I'm looking for. A modular implementation of the BOLT specifications. And it's first use case, research, is what I need it for.
I'm also in the position to invest considerable time in this repository, if it actually fits my needs, and it seems that it does.

That being said, after playing around with it I have some questions about the current status.

This is what I got so far:

- the daemon didn't work for me. It directly exits with an error.
- If I run the tests, 21 test cases of the noise package fail, after which the test suite exits.

After that I tried to start smaller, so I tried to get the NoiseSocket server running and that worked! But when I connect with a client, I get an error. I do get a socket connection, but the handshake fails. It fails with this error:

```
ReferenceError: next is not defined
    at Cipher._update (C:\Repositories\lntools\packages\lntools-noise\node_modules\chacha\aead.js:39:5)
    at CipherBase.update (C:\Repositories\lntools\packages\lntools-noise\node_modules\chacha\cipherbase.js:40:22)
    at ccpEncrypt (C:\Repositories\lntools\packages\lntools-noise\node_modules\@lntools\crypto\lib\crypto.js:133:20)
    at NoiseState.initiatorAct1 (C:\Repositories\lntools\packages\lntools-noise\lib\noise-state.js:185:13)
    at NoiseSocket._initiateHandshake (C:\Repositories\lntools\packages\lntools-noise\lib\noise-socket.js:261:30)
    at NoiseSocket._onConnected (C:\Repositories\lntools\packages\lntools-noise\lib\noise-socket.js:192:14)
    at Socket.emit (events.js:194:15)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:10)
Emitted 'error' event at:
    at Socket.NoiseSocket._socket.on.err (C:\Repositories\lntools\packages\lntools-noise\lib\noise-socket.js:158:42)
    at Socket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

I think that's because in `noise-state.js` line 185 an zero sized Buffer is passed as `plaintext` argument, and it seems that ChaCha doesn't like that.

So that's where I am right now. My question is: @bmancini55 , did you get things to run so far? Am I doing something wrong? Do you have a local version with stuff that's still to be committed?

Thanks in advance for any help or pointers.

**Comment by bmancini55** [comment]
Thank you for writing and checking out the repo!  It would be great to have assistance on these tools!  What areas of research are you working on?

The repo is being actively worked on a few days per week.  I'm not able to dedicate full time to it at the moment (I would certainly love to though). Efforts over the last ~month were spent on metawork to get the repository easier to work with as a modular system. Some of those changes include:

* breaking out from a single project into a monorepo with Lerna
* adding CI with CircleCI across all projects
* adding code coverage with codecov.io across all projects
* improving and standardizing JSDoc/TSDoc across all projects
* adding build-time JavaScript type checking with TypeScript (note we're not using TypeScript but checking using JSDoc/TSDoc definitions)

I've been the primary person working on this repo for now, so the master branch is rather unstable until these changes are finalized. I also have a handful of projects locally that haven't been pushed to the repo since they're not in the Lerna format yet.

The release modules of `@lntools/noise` and `@lntools/invoice` on NPM should be stable. 

On the issues you're encountering:

* `src/daemon` is woefully out of date. I can clean that up shortly and get it checked into master. It will likely be replaced by more specific tools as the project progresses. It was originally just a testing ground for connecting to a Bitcoin testnet node and ensuring connection state can be maintained.
* I will take a look into `noise-state.js:L185` and see what might be going on.  CI builds are passing for `noise`, so I wonder if there is some compatibility issue with Windows (I'm running on Mac and CI runs in Linux). Also I just pulled a fresh version of the repo, ran NPM install, and was able to get a simple test application below running. (Note this file was dropped to `packages/lntools-noise/test.js`)

```javascript
// packages/lntools-noise/test.js
// @ts-check

/* eslint-disable no-console */
const crypto = require('@lntools/crypto');

const winston = require('winston');
winston.configure({
  level: 'info',
  transports: [new winston.transports.Console()],
  format: winston.format.combine(
    winston.format.splat(),
    winston.format.colorize(),
    winston.format.simple()
  ),
});

const { connect, createServer } = require('./lib');

async function run() {
  startServer();
  createClient();
}

run().catch(console.error);

function startServer() {
  let ls = Buffer.from('2121212121212121212121212121212121212121212121212121212121212121', 'hex');
  let esFactory = () => Buffer.from('2222222222222222222222222222222222222222222222222222222222222222', 'hex'); // prettier-ignore
  let msg = 0;

  // fire up the server
  let server = createServer({ ls, esFactory }, socket => {
    winston.info('server socket open');

    socket.on('connect', () => {
      winston.info('server socked connected');
      socket.write(Buffer.from('hello ' + ++msg));
    });

    socket.on('data', d => {
      winston.info('server received: ' + d.toString());
      socket.write(Buffer.from('hello ' + ++msg));
    });
  });

  server.listen({ port: 9735, host: '127.0.0.1' });
}

function createClient() {
  let ls = Buffer.from('1111111111111111111111111111111111111111111111111111111111111111', 'hex');
  let rpk = crypto.getPublicKey(
    Buffer.from('2121212121212121212121212121212121212121212121212121212121212121', 'hex')
  );
  let es = Buffer.from('1212121212121212121212121212121212121212121212121212121212121212', 'hex');
  let msg = 0;

  let socket = connect({
    ls,
    es,
    rpk,
    host: 'localhost',
    port: 9735,
  });

  socket.on('connect', () => {
    winston.info('client connected');
  });

  socket.on('data', data => {
    winston.info('client received: ' + data.toString());
    setTimeout(() => {
      socket.write(Buffer.from('hello ' + ++msg));
    }, 250);
  });
}

```

**Comment by gijswijs** [comment]
Thanks for getting back to me on such a short notice. I'll look into things
in more detail tomorrow and get back to you.

I'm based in Kuala Lumpur and it 22:30 over here, so I'm calling it a day.

**Comment by bmancini55** [comment]
Sounds good.  The `chacha` library uses `chacha-native` under the covers. My install built the C++ library, so there could be a difference between the c++ and pure js implementations.

**Comment by bmancini55** [comment]
@gijswijs I just pushed a few changes to update daemon. You will need to run `npm run link` in the root of lntools prior to running it, otherwise daemon won't be able to find the `@lntools/wire` package.

**Comment by gijswijs** [comment]
@bmancini55 I just made a small test with the release module of @lntools/noise on NPM.
When installing, I saw that node-gyp defaulted to a pure JS implementation. I think I missed that the other day.

I installed the Visual C++ Build Tools in VS2015, and set the NPM config key `npm config set msvs_version 2015`

After that the install of @lntools/noise didn't default to the pure JS version of chacha-native.
And consequently the client-server example ran smoothly.

Just to be sure I also tested the pure JS build with the release module of @lntools/noise on NPM and that gave the same errors I mentioned earlier. So it's save to say that the C++ build and the pure JS build aren't exact ports of each other.

Now that we got that out of the way, I'm very eager to go ahead and try the daemon you just updated.

To answer your earlier question: My research is on privacy and second layer solutions, mainly Lightning. It's PhD research so I'll be doing this for the next three years if everything goes all right. What I try to do is test if the theoretical improvements of 2nd layer solutions with regards to privacy hold up in practice.

For now I'm interested in the opening and closing of LN channels. For this I need to scan the gossip messages to see when channels open, and then check the blockchain to see when they close and whether this was mutual or unilateral (or revoked).

As such, at his moment, I'm interested in BOLT 7, and obviously BOLT 1 and 8 because without those 2 you got nothing.

**Comment by gijswijs** [comment]
> I just pushed a few changes to update daemon

Are you sure? I don't see any new commits yet.

**Comment by bmancini55** [comment]
@gijswijs ah sorry. Ran out the door too quickly. You should be good now.

**Comment by gijswijs** [comment]
I think we got something running!

```
info: peer initialized {"type":16,"globalFeatures":"0","localFeatures":"82"}
```

I'm really happy with this. If you want, I can close this issue, because the answer is that the current status of this repository is pretty much reflected the README.md.

But if you have some pointers on where you want this repository to go, let me know, and I'll see if I can help out.

**Comment by bmancini55** [comment]
I'm glad you were able to get the native version working! I'm going to change that package to use `chacha-native` directly to prevent any future issues.

Also we can update master README.md file to include your notes for windows users.

Your research sounds great and aligns nicely with our initial goals! 

One of Altangent's main projects is the data analysis platform www.blocktap.io.  We've been collecting statistics on the Lightning Network for a while and working on a variety of tools (https://graph.lndexplorer.com/ for one). The first goals of lntools is collection of p2p message and on-chain monitoring so that we can achieve more granular lightning data for blocktap.  

I have a few projects that need to be merged into lntools:
* A bitcoin project that is a functional library for building and parsing transactions
* A graph project that constructs a graph from network messages with a similar structure to LND's graph API
*  A chain monitor that uses the bitcoind connector package and emits transactions that match user specified patterns. The pattern matching part still needs to developed.

The wire projects currently needs a good bit of work.  It is the code for the LN wire protocol and  currently includes the peer wrapper for the noise socket, ping/pong state machine, and partially implemented parsing/building messages.

**Comment by gijswijs** [comment]
I've been thinking about this. I'd love to help. Let me be honest. My development skills are rusty, to say the least. I've been working in my own company for the past 17 years, and development was less and less part of my job. Only recently the opportunity arose to do my PhD, and I grabbed it. So I've been brushing up on my javascript/node skills (and my C# skills, but that's not really relevant for this repo)
Still, I think I can be of help.
Those projects that you talk about? Are those repo's already on github? How are you monitoring the chain? Do you use ZeroMQ for that? In What way do you want to parse bitcoin transactions? Do you use a GDB for your graph needs? If so, which one?
So many questions...

**Comment by bmancini55** [comment]
Any help would be appreciated!  

Most of the code is either in local branches (going to push them to GitHub branches today) or scattered throughout this repo: https://github.com/bmancini55/crypto-playpen.   

For connecting to bitcoind, the package lntools-bitcoind already exists. It is just an RPC and ZeroMQ wrapper and some TypeScript type definitions.

Packages that will be pushed today: 
  - `bitcoin` - a functional library for building and parsing transactions.  A lot of the code was inspired by bitcoinjs-lib, but I've been playing with more flexible transaction building. This code exists in https://github.com/bmancini55/crypto-playpen/tree/master/bitcoin2/src.  I've only pulled in the code for tx parsing so far.  The tx building code is next on my list.
  - `chainmon` - combines `bitcoin` and `bitcoind` packages to listen to ZeroMQ stream and parse txs real time.  As a mentioned above, the goal of this lib is to allow subscribing to txids or transaction patterns and emitting events when a match is seen.  
  - `graph` - manages p2p messages to construct a valid graph.  p2p message traffic has a few complexities that require cascading state to validate that messages are authentic and that a complete graph picture can be built.  I have the basic logic working for constructing channels and nodes in a format compatible with LND's graph API results.  

The initial goal will be building an in-memory graph using the graph package. Additional graph traversal/routing logic could go there as well.  At some point this could be serialized to disk or pushed to a GDB pretty easily.

**Comment by gijswijs** [comment]
Hey man,

You are on a rampage lately with this repository. Great to see.

**Comment by bmancini55** [comment]
Thanks! Been trying to get all the gossip code buttoned up. Almost there. More examples and better documentation are on the list as well.
